**Student Performance Indicator**

Life Cycle of Machine Learning Project

 Understanding the problem statement
 Data Collection
 Data Checks to perform
 Exploratory data analysis
 Date Pre-Processing
 Model Training
 Choose best model


1) Problem Statement

This project understands how the student's performance (test score) is affected by other variables such as Gender, Ethnicity, Parental level of education, Lunch and Test preparation course.


2) Data Collection \
Data Source - https://www.kaggle.com/datasets/spscientist/students-performance-in-exams?datasetId=74977

The data consist of 8 columns and 1000 rows

2.1 Import Data and Requried Packages



In [7]:
print("Hello")

Hello


In [11]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

Import the CSV Data as Pandas DataFrame

In [15]:
df=pd.read_csv('data\StudentsPerformance.csv')

Show Top 5 Records

In [16]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

Shape of the dataset

In [17]:
df.shape

(1000, 8)

2.2 Dataset Information \

gender : sex of students => (Male/Feamale) \
race/ethnicity : ethinicity of students => (Group A,B,C,D,E) \
parental level of education : parents final education => (bachelor's degree, some college, master's degree, associate's degree, high school) \
lunch : having lunch before test (standard or free/reduced) \
test preparation course : complete or not complete before test \
math score \
reading score \
writing score

3. Data Checks to perform

Check missing values \
check duplicates \
Check data type \
check the number of unique values of each column \
check statistics of dataset \
check various categories present in the different categorical column

3.1 Check Missing values

In [18]:
df.isna().sum()

gender                         0
race/ethnicity                 0
parental level of education    0
lunch                          0
test preparation course        0
math score                     0
reading score                  0
writing score                  0
dtype: int64

There are no missing values in the data set

3.2 Check Duplicates

In [19]:
df.duplicated().sum()

0

There are no duplicates values in the data set

3.3 Check data types

In [20]:
#check null and datatype
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


3.4 Checking the number of unique values of each column

In [21]:
df.nunique()

gender                          2
race/ethnicity                  5
parental level of education     6
lunch                           2
test preparation course         2
math score                     81
reading score                  72
writing score                  77
dtype: int64

3.5 Check statistics of dataset

In [22]:
df.describe()

math score  reading score  writing score
count  1000.00000    1000.000000    1000.000000
mean     66.08900      69.169000      68.054000
std      15.16308      14.600192      15.195657
min       0.00000      17.000000      10.000000
25%      57.00000      59.000000      57.750000
50%      66.00000      70.000000      69.000000
75%      77.00000      79.000000      79.000000
max     100.00000     100.000000     100.000000

**Insight** \
From above description of numerical data, all means are very close to each other - between 66 and 68.05; \
All standard deviations are also close - between 14.6 and 15.19; \
While there is a minimum score 0 for math, for writing minimum is much higher = 10 and for reading might higher = 17

3.7 Exploring Data

In [23]:
df.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

In [28]:
print("categories in 'gender' variable:  ",end=" ")
print(df['gender'].unique())

print("categories in 'race_ethnicity' variable:  ",end=" ")
print(df['race/ethnicity'].unique())

print("categories in 'parental level of education' variable:  ",end=" ")
print(df['parental level of education'].unique())

print("categories in 'lunch' variable:  ",end=" ")
print(df['lunch'].unique())

print("categories in 'test preparation course' variable:  ",end=" ")
print(df['test preparation course'].unique())


categories in 'gender' variable:   ['female' 'male']
categories in 'race_ethnicity' variable:   ['group B' 'group C' 'group A' 'group D' 'group E']
categories in 'parental level of education' variable:   ["bachelor's degree" 'some college' "master's degree" "associate's degree"
 'high school' 'some high school']
categories in 'lunch' variable:   ['standard' 'free/reduced']
categories in 'test preparation course' variable:   ['none' 'completed']


In [29]:
#define numerical & categorical columns

numeric_feature = [i for i in df.columns if df[i].dtype != 'O']
categorical_feature = [i for i in df.columns if df[i].dtype == 'O']

In [32]:
#print columns
print(len(numeric_feature),"numeric features: ", numeric_feature)
print(len(categorical_feature),"categorical features: ",categorical_feature)

3 numeric features:  ['math score', 'reading score', 'writing score']
5 categorical features:  ['gender', 'race/ethnicity', 'parental level of education', 'lunch', 'test preparation course']


3.8 Adding columns for "Total Score" and "Average"